rig/machine_control/scp_connection.py

```diff
@@ -79,65 +125,167 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
+
+        # While there are packets in the queue or packets for which we are
+        # still awaiting returns then continue to loop.
+        while queued_packets or outstanding_packets:
+            # If there are fewer outstanding packets than the window can take
+            # and we still might have packets left to send then transmit a
+            # packet and add it to the list of outstanding packets.
+            if len(outstanding_packets) < window_size and queued_packets:
+                try:
+                    args = next(parameters_and_callbacks)
+                except StopIteration:
+                    queued_packets = False
+
+                if queued_packets:
+                    # If we extracted a new packet to send then create a new
+                    # outstanding packet and transmit it.
+                    seq = next(self.seq)
+                    while seq in outstanding_packets:
+                        # The seq should rarely be already taken, it normally
+                        # means that one packet is taking such a long time to
+                        # send that the sequence has wrapped around.  It's not
+                        # a problem provided that we don't reuse the number.
+                        seq = next(self.seq)
+
+                    # Construct the packet that we'll be sending
+                    packet = SCPPacket(
+                        reply_expected=True, tag=0xff, dest_port=0,
+                        dest_cpu=args.p, src_port=7, src_cpu=31,
+                        dest_x=args.x, dest_y=args.y, src_x=0, src_y=0,
+                        cmd_rc=args.cmd, seq=seq,
+                        arg1=args.arg1, arg2=args.arg2, arg3=args.arg3,
+                        data=args.data
                     )
-                )
 
-            if scp.seq == self._seq:
-                # The packet is the acknowledgement.  Increment the
-                # sequence indicator and return the packet.
-                self._seq ^= 1
-                return scp
+                    # Create a reference to this packet so that we know we're
+                    # expecting a response for it and can retransmit it if
+                    # necessary.
+                    outstanding_packets[seq] = TransmittedPacket(
+                        args.callback, packet.bytestring, args.expected_args,
+                        args.timeout
+                    )
 
-        # The packet we transmitted wasn't acknowledged.
-        raise TimeoutError(
-            "Exceeded {} tries when trying to transmit packet.".format(
-                self.n_tries)
-        )
+                    # Actually send the packet
+                    self.sock.send(packet.bytestring)
 
-    def read(self, buffer_size, x, y, p, address, length_bytes):
+            # Listen on the socket for an acknowledgement packet, there not be
+            # one.
+            try:
+                ack = self.sock.recv(receive_length)
```

Comment:
Ah... a busy loop. I think I'd push for swapping this out for a select+timeout (with timeout set to the next timeout to occur) before we merge this branch in. Shouldn't be a big change (I don't think!) but should save battery-users and multi-taskers some time!